src/eos-updater-apply.c

```diff
@@ -73,41 +77,506 @@ get_test_osname (void)
+                                cancellable,
+                                error))
+    {
+      eos_updater_remove_recursive (checkout_directory, NULL);
+      return FALSE;
+    }
+
+  autoinstall_flatpaks = list_flatpak_refs_from_d_directory (checkout_directory,
+                                                             cancellable,
+                                                             error);
+  eos_updater_remove_recursive (checkout_directory, NULL);
+
+  if (!ostree_repo_checkout_gc (repo, cancellable, error))
+    return FALSE;
+
+  *out_flatpaks_this_commit_wants = g_steal_pointer (&autoinstall_flatpaks);
+  return TRUE;
+}
+
+static FlatpakInstallation *
+get_flatpak_installation (GCancellable *cancellable, GError **error)
+{
+  const gchar *override_path = g_getenv ("EOS_UPDATER_TEST_FLATPAK_INSTALLATION_DIR");
+
+  if (override_path)
+    {
+      g_autoptr(GFile) override = g_file_new_for_path (override_path);
+      return flatpak_installation_new_for_path (override, TRUE, cancellable, error);
+    }
+
+  return flatpak_installation_new_system (cancellable, error);
+}
+
+static guint
+flatpak_ref_hash (gconstpointer ref)
+{
+  return g_str_hash (flatpak_ref_get_name ((FlatpakRef *) ref));
+}
+
+static gboolean
+flatpak_ref_equal (gconstpointer a, gconstpointer b)
+{
+  return g_str_equal (flatpak_ref_get_name ((FlatpakRef *) a),
+                      flatpak_ref_get_name ((FlatpakRef *) b));
+}
+
+static GHashTable *
+determine_set_of_installed_applications (FlatpakInstallation *installation,
+                                         GCancellable        *cancellable,
+                                         GError              **error)
+{
+  g_autoptr(GPtrArray) refs = flatpak_installation_list_installed_refs (installation,
+                                                                        cancellable,
+                                                                        error);
+  g_autoptr(GHashTable) set = NULL;
+  gsize i = 0;
+
+  if (!refs)
+    return NULL;
+
+  set = g_hash_table_new_full (flatpak_ref_hash,
+                               flatpak_ref_equal,
+                               g_object_unref,
+                               NULL);
+
+  for (; i < refs->len; ++i)
+    g_hash_table_add (set, g_ptr_array_index (refs, i));
+
+  return g_steal_pointer (&set);
+}
+
+static GPtrArray *
+resolve_flatpaks_to_deploy (GPtrArray    *flatpaks_this_commit_wants,
+                            GCancellable *cancellable,
+                            GError       **error)
+{
+  g_autoptr(FlatpakInstallation) installation = get_flatpak_installation (cancellable, error);
+  g_autoptr(GPtrArray) resolved_flatpaks = g_ptr_array_new_with_free_func (g_object_unref);
+  g_autoptr(GHashTable) explicitly_removed_flatpaks = NULL;
+  g_autoptr(GHashTable) already_installed_flatpaks = NULL;
+  gsize i;
+
+  already_installed_flatpaks = determine_set_of_installed_applications (installation, cancellable, error);
```

Comment:
This seems like a pretty inefficient way to reach the goal stated below (filter out from a list flatpaks that are already installed). Instead of taking the whole array, turning it into a hash table, and then using `g_hash_table_contains()`,  you could instead just use e.g. http://flatpak.org/flatpak/reference/html/FlatpakInstallation.html#flatpak-installation-get-current-installed-app and check directly if the app is installed.

Reply:
Yeah - I had it done this way in the past so that we could merge the list of both installed and uninstalled apps and check against both at the same time, but obviously that's not relevant anymore. I'll just use the API directly...